### PR TITLE
integrate new logger framework

### PR DESCRIPTION
**Pull request by Reidmcc**
This PR continues https://github.com/interstellar/kelp/issues/9

I've integrated the new logger into the various structs and `make...` functions. It does **not** compile, pending some questions.

I also revised some functions in lower level packages that should return errors instead of doing their own logging, which is what the majority of the functions in those packages do. Namely in `model/assets.go`, `model/botkey.go` and `support/utils/configs.go`. I did not yet go through and revise every call to these functions, I wanted to confirm that this approach makes sense first. Hence the not compiling.

I wanted to do the same thing for `support/utils/functions.go` but its functions are so nested that changing from self-logging to outputting loggable errors in some of the functions would cascade into other functions having to check multiple possible duplicate errors and then output those yet again. All of which seems massively bloated. `PriceAsFloat` is a big offender here as it is called multiple times in functions like `Less`. What do you think?

What should we do with log lines followed by `panic`, e.g. `support/networking/server.go` lines 98-99. Should they be switched to `panic(fmt.Errorf(...)` as in `cmd/exchanges.go` line 31?

You'll see that my IDE did a couple weird things. It added a blank line after the import of `logger` in most files every time I saved the file. Is that right for some reason? It also aliased a couple imports, e.g. `krakenapi "github.com/Beldur/kraken-go-api-client"`. The code uses those aliases, so that seems reasonable.

This PR is showing the previously merged changes from https://github.com/interstellar/kelp/pull/88 as new; not sure what that's about.

File status details:
````
Files that need framework conversion:
	support/networking
		server.go - pending decision on what to do with panics
	suport/utils
		functions.go - pending decision on what to do with all the functions error cascading


Files ready for find-replace:
	api/
		priceFeed.go
	plugins/
		exchangeFeed.go
		ccxtExchange.go
		deleteSideStrategy.go
		fillLogger.go
		fillTracker.go
		intervalTimeController.go
		krakenExchange.go
		mirrorStrategy.go
		sdex.go
		sellSideStrategy.go
		staticSpreadLevelProvider.go
	support/monitoring/
		metricsEndpoint.go
		pagerDuty.go
	support/sdk
		ccxt.go
	support/utils
		configs.go
	terminator/
		terminator.go
	trader/
		trader.go

Files fully converted to new logger framework, i.e. don't have log output lines to replace:
	api/
		assets.go
	model/
		assets.go
		botKey.go
	cmd/
		trade.go
	plugins/
		cmcFeed.go
		fiatFeed.go
		fixedFeed.go
		balancedLevelProvider.go

Files that don't need converting (_test and non-.go files not listed):

main.go
	accounting/pnl/
		pnl.go (not actually Kelp)
	cmd/
		exchanges.go (wouldn't log this)
		root.go
		strategies.go (wouldn't log this)
		version.go
	api/
		alert.go
		exchange.go
		level.go
		strategy.go
		timeController.go
	model/
		dates.go
		number.go
		orderbook.go
		tradingPair.go
	plugins/
		balancedStrategy.go
		buysellStrategy.go
		composeStrategy.go
		deleteStrategy.go
		factory.go
		priceFeed.go
		sellStrategy.go
	support/monitoring
		factory.go
		metrics.go
		metricsRecorder.go
	support/networking
		endpoint.go
		network.go
		parser.go
	support/utils
		config.go
	terminator/
		config.go
	trader/
		config.go
````

### CODE REVIEW

File: cmd/trade.go [hunk 1]
```diff
@@ -45,10 +46,10 @@ func hiddenFlag(flag string) {
 	}
 }
 
-func logPanic() {
+func logPanic(l logger.Logger) {
 	if r := recover(); r != nil {
 		st := debug.Stack()
-		log.Printf("PANIC!! recovered to log it in the file\npanic: %v\n\n%s\n", r, string(st))
+		l.Errorf("PANIC!! recovered to log it in the file\npanic: %v\n\n%s\n", r, string(st))
```
Comment (nikhilsaraf) [inline cmd/trade.go]: @Reidmcc  can you rebase off the latest master (it already contains some of these changes)

File: api/priceFeed.go [hunk 2]
```diff
@@ -11,6 +15,7 @@ type PriceFeed interface {
 type FeedPair struct {
 	FeedA PriceFeed
 	FeedB PriceFeed
+	l     logger.Logger
```
Comment (nikhilsaraf) [inline api/priceFeed.go]: this looks like it's unused? -- better to combine it with the factory method that creates this otherwise it's a little harder to review :)

File: model/botKey.go [hunk 3]
```diff
@@ -30,10 +30,10 @@ func (b BotKey) String() string {
 }
 
 // MakeSortedBotKey makes a BotKey by sorting the passed in assets
-func MakeSortedBotKey(assetA horizon.Asset, assetB horizon.Asset) *BotKey {
+func MakeSortedBotKey(assetA horizon.Asset, assetB horizon.Asset) (*BotKey, error) {
 	var assetBaseCode, assetBaseIssuer, assetQuoteCode, assetQuoteIssuer string
 	if assetA.Type == native && assetB.Type == native {
-		log.Fatal("invalid asset types, both cannot be native")
+		return nil, fmt.Errorf("invalid asset types, both cannot be native")
```
Comment (nikhilsaraf) [inline model/botKey.go]: let's not fix any place where we are using log.Fatal right now. this PR should be a simple translation of old format to new.

File: plugins/balancedLevelProvider.go [hunk 4]
```diff
@@ -59,6 +62,7 @@ func makeBalancedLevelProvider(
 	virtualBalanceQuote float64,
 	orderConstraints *model.OrderConstraints,
 ) api.LevelProvider {
+	l := logger.MakeBasicLogger()
```
Comment (nikhilsaraf) [inline plugins/balancedLevelProvider.go]: this should reuse the existing logger that is passed in. we should have only one call to MakeBasicLogger throughout Kelp

File: model/assets.go [hunk 5]
```diff
@@ -82,12 +81,12 @@ func (c AssetConverter) FromString(s string) (Asset, error) {
 }
 
 // MustFromString converts from a string to an asset, failing on errors
-func (c AssetConverter) MustFromString(s string) Asset {
+func (c AssetConverter) MustFromString(s string) (Asset, error) {
 	a, e := c.FromString(s)
 	if e != nil {
-		log.Fatal(fmt.Errorf("exiting on an error-enforced asset conversion: %s", e))
+		return "", fmt.Errorf("exiting on an error-enforced asset conversion: %s", e)
```
Comment (nikhilsaraf) [inline model/assets.go]: this method should be a fatal -- MustFromString forces you to panic by definition.
let's not make any conversions from fatal to returning an error in this PR. 

the logger should be injected into the AssetConverter. Let's try that for now and see how big the change is

File: plugins/ccxtExchange.go [hunk 6]
```diff
@@ -32,6 +35,7 @@ func makeCcxtExchange(
 	apiKeys []api.ExchangeAPIKey,
 	simMode bool,
 ) (api.Exchange, error) {
+	l := logger.MakeBasicLogger()
```
Comment (nikhilsaraf) [inline plugins/ccxtExchange.go]: here too, don't instantiate logger here but take it in as a param to the make function

File: plugins/fillLogger.go [hunk 7]
```diff
@@ -3,18 +3,25 @@ package plugins
...
+type FillLogger struct {
+	l logger.Logger
+}
 
 var _ api.FillHandler = &FillLogger{}
 
 // MakeFillLogger is a factory method
 func MakeFillLogger() api.FillHandler {
-	return &FillLogger{}
+	l := logger.MakeBasicLogger()
```
Comment (nikhilsaraf) [inline plugins/fillLogger.go]: here too, don't instantiate a new logger -- take in an argument in MakeFillLogger and use that to set `fillLogger.l`

File: plugins/intervalTimeController.go [hunk 8]
```diff
@@ -13,15 +15,18 @@ type IntervalTimeController struct {
 	tickInterval       time.Duration
 	maxTickDelayMillis int64
 	randGen            *rand.Rand
+	l                  logger.Logger
 }
 
 // MakeIntervalTimeController is a factory method
 func MakeIntervalTimeController(tickInterval time.Duration, maxTickDelayMillis int64) api.TimeController {
+	l := logger.MakeBasicLogger()
```
Comment (nikhilsaraf) [inline plugins/intervalTimeController.go]: @Reidmcc  reuse of basic logger object applies everywhere.

File: plugins/krakenExchange.go [hunk 9]
```diff
@@ -9,7 +9,9 @@ import (
 	"strconv"
 	"strings"
 
-	"github.com/Beldur/kraken-go-api-client"
+	"github.com/interstellar/kelp/support/logger"
+
+	krakenapi "github.com/Beldur/kraken-go-api-client"
```
Comment (nikhilsaraf) [inline plugins/krakenExchange.go]: this import should not have changed: `krakenapi "github.com/Beldur/kraken-go-api-client"`

File: support/logger/basicLogger.go [hunk 10]
```diff
@@ -0,0 +1,35 @@
+package logger
+
+import "log"
+
+// basicLogger is a standard logger
+type basicLogger struct {
```
Comment (nikhilsaraf) [inline support/logger/basicLogger.go]: this should not be a part of this PR since it's already in the master repo

File: support/logger/logger.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+func Fatal(l Logger, e error) {
+	l.Info("")
+	l.Errorf("%s", e)
+	os.Exit(1)
+}
+
+/*
+Notes:
+	What to do with log lines followed by panic? Should those just go to logger.Fatal and exit?
+	What to do with utils.PriceAsFloat? If you have it return an error many of the referencing functions have to parse multiple errors.
```
Comment (nikhilsaraf) [inline support/logger/logger.go]: functions can have a global variable called `utilsLogger` which we set by calling `utils.SetLogger()` in `trade.go`:
```golang
func SetLogger(l logger.Logger) {
    if utilsLogger != nil {
        // protect against resetting of logger
        logger.Fatal(utilsLogger, fmt.Errorf("cannot reset the logger since it's non-nil"))
    }
    utilsLogger = l
}
```

utils.PriceAsFloat can use utilsLogger for now. This will simplify logging for this file for now.

File: support/logger/logger.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+// Fatal is a convenience method that can be used with any Logger to log a fatal error
+func Fatal(l Logger, e error) {
+	l.Info("")
+	l.Errorf("%s", e)
+	os.Exit(1)
+}
+
+/*
+Notes:
+	What to do with log lines followed by panic? Should those just go to logger.Fatal and exit?
```
Comment (nikhilsaraf) [inline support/logger/logger.go]: we should leave a panic unchanged. It will be handled automatically in logPanic()

File: support/logger/logger.go [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+	// error messages, indicates to the logger that these messages can be handled differently (different color, special format, email alerts, etc.). The type of logger will determine what to do with these messages. The logger should NOT panic on these messages. User needs to add a \n if they want a newline after the log entry.
+	Errorf(msg string, args ...interface{})
+}
+
+// Fatal is a convenience method that can be used with any Logger to log a fatal error
+func Fatal(l Logger, e error) {
+	l.Info("")
+	l.Errorf("%s", e)
+	os.Exit(1)
```
Comment (nikhilsaraf) [inline support/logger/logger.go]: @Reidmcc just realized this, it's very important that we change the `l.Errorf("%s", e)` followed by `os.Exit(1)` to a single `panic(e)` call.
if we call os.Exit here then it will not be written to the file that we log too!

File: support/logger/logger.go [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+/*
+Notes:
+	What to do with log lines followed by panic? Should those just go to logger.Fatal and exit?
+	What to do with utils.PriceAsFloat? If you have it return an error many of the referencing functions have to parse multiple errors.
+		It's really bloated and there's no struct to drop a logger into
+		Passing a logger into each function here seems ridiculous
```
Comment (nikhilsaraf) [inline support/logger/logger.go]: you're right, we should not pass a logger into each function. The logger object should **only** be on structs/objects, not passed into functions (two exceptions: logPanic and validateCliParams)

File: support/logger/logger.go [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+		Overall this file gets nasty with the new logger framework
+
+Files that need framework conversion:
```
Comment (nikhilsaraf) [inline support/logger/logger.go]: please keep framework usage in a separate PR (blocked on this PR for now)

File: support/logger/logger.go [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+	support/networking
+		server.go - pending decision on what to do with panics
+	suport/utils
+		functions.go - pendimg decision on what to do with all the functions error cascading
```
Comment (nikhilsaraf) [inline support/logger/logger.go]: there should not be any cascading errors since we should not be making any changes to the error flow with this PR

File: support/monitoring/pagerDuty.go [hunk 17]
```diff
@@ -4,20 +4,25 @@ import (
 	"fmt"
 	"log"
 
-	"github.com/PagerDuty/go-pagerduty"
+	"github.com/interstellar/kelp/support/logger"
+
+	pagerduty "github.com/PagerDuty/go-pagerduty"
```
Comment (nikhilsaraf) [inline support/monitoring/pagerDuty.go]: incorrect import format, should be unchanged for pagerduty

File: support/utils/configs.go [hunk 18]
```diff
@@ -6,10 +6,12 @@ import (
 	"log"
 	"reflect"
 	"strings"
+
+	"github.com/interstellar/kelp/support/logger"
 )
 
 // CheckConfigError checks configs for errors, crashes app if there's an error
-func CheckConfigError(cfg fmt.Stringer, e error, filename string) {
+func CheckConfigError(l logger.Logger, cfg fmt.Stringer, e error, filename string) {
```
Comment (nikhilsaraf) [inline support/utils/configs.go]: this can use utils.utilsLogger instead of threading through these functions (utilsLogger will be private to utils package with the global variable usage describe above)

**Comment by nikhilsaraf** [comment]
@Reidmcc also please rebase this PR off the master branch so it excludes changes that are already in master.

File: cmd/trade.go (hunk 1 above)
Comment (Reidmcc) [reply cmd/trade.go]: yep

File: model/assets.go (hunk 5 above)
Comment (Reidmcc) [reply model/assets.go]: ok

File: model/botKey.go (hunk 3 above)
Comment (Reidmcc) [reply model/botKey.go]: ok

File: support/logger/logger.go (hunk 13 above)
Comment (Reidmcc) [reply support/logger/logger.go]: Ah, makes sense

File: plugins/krakenExchange.go (hunk 9 above)
Comment (Reidmcc) [reply plugins/krakenExchange.go]: how does the `apis                     []*krakenapi.KrakenApi` work in the KrakenExchange struct work if there's no `krakenapi` import?

File: api/priceFeed.go (hunk 2 above)
Comment (Reidmcc) [reply api/priceFeed.go]: there's a log call on line 35 that uses it; that does seem like a weird place to log though

**Comment by Reidmcc** [comment]
@nikhilsaraf what do you want to do with the `terminate` command logging? Leave it alone for now?

**Comment by Reidmcc** [comment]
@nikhilsaraf ugh, I screwed up this branch git-wise. After the rebase github required that I pull before pushing and it re-merged old changes that are now written as more commits. The last normal commit is https://github.com/interstellar/kelp/pull/99/commits/2527e0ce6d050d84ada584e6ccafab6d3d8fb591, labeled "more log passing" above, the remainder are junk merge artifacts. Well, that and giving `pnl.go` a logger to pass.

**Comment by Reidmcc** [comment]
@nikhilsaraf I don't know what is going on with the `krakenExchange_test` failing at `TestGetTickerPrice` on the Travis check

File: support/utils/configs.go (hunk 18 above)
Comment (Reidmcc) [reply support/utils/configs.go]: done

File: plugins/balancedLevelProvider.go (hunk 4 above)
Comment (Reidmcc) [reply plugins/balancedLevelProvider.go]: done

File: plugins/ccxtExchange.go (hunk 6 above)
Comment (Reidmcc) [reply plugins/ccxtExchange.go]: done

File: support/monitoring/pagerDuty.go (hunk 17 above)
Comment (Reidmcc) [reply support/monitoring/pagerDuty.go]: fixed

**Comment by Reidmcc** [comment]
@nikhilsaraf  changes made, bugs fixed, ready for review

**Comment by Reidmcc** [comment]
Hold up, some changes I was pretty sure I pushed aren't right here. Not actually ready for review.

File: cmd/terminate.go [hunk 19]
```diff
@@ -24,6 +25,7 @@ func init() {
 	configPath := terminateCmd.Flags().StringP("conf", "c", "./terminator.cfg", "service's basic config file path")
 
 	terminateCmd.Run = func(ccmd *cobra.Command, args []string) {
+		l := logger.MakeBasicLogger()
```
Comment (Reidmcc) [inline cmd/terminate.go]: can't use the trade.go logger here since its a different command

**Comment by Reidmcc** [comment]
@nikhilsaraf the revisions are ready for review. Travis seems to be stuck for this PR

File: plugins/krakenExchange.go (hunk 9 above)
Comment (nikhilsaraf) [reply plugins/krakenExchange.go]: we import it without the alias, which seems to work. if you can't get that to work then the alias is ok

File: support/utils/functions.go [hunk 20]
```diff
@@ -237,6 +240,14 @@ func CheckedString(v interface{}) string {
 	return fmt.Sprintf("%v", v)
 }
 
+// SetLogger sets the global utilsLogger var
+func SetLogger(l logger.Logger) {
```
Comment (nikhilsaraf) [inline support/utils/functions.go]: 👍

File: support/logger/logger.go [hunk 21]
```diff
@@ -21,5 +19,5 @@ type Logger interface {
 func Fatal(l Logger, e error) {
 	l.Info("")
 	l.Errorf("%s", e)
-	os.Exit(1)
+	panic(e)
```
Comment (nikhilsaraf) [inline support/logger/logger.go]: thanks!

**Comment by Reidmcc** [comment]
@nikhilsaraf  This pr is ridiculously out of date, we should probably close it.

**Comment by nikhilsaraf** [comment]
yeah, probably makes sense to close out, there's also merge conflicts because of the change from interstellar/kelp -> stellar/kelp.
